gzip/gzip.go

```diff
@@ -67,44 +69,45 @@ func (c CompressionCodec) Code() int8 {
 	return Code
 }
 
-// Encode implements the kafka.CompressionCodec interface.
-func (c CompressionCodec) Encode(src []byte) ([]byte, error) {
-	buf := bytes.Buffer{}
-	buf.Grow(len(src)) // guess a size to avoid repeat allocations.
-	writer := writerPool.Get().(*gzip.Writer)
-	writer.Reset(&buf)
-
-	_, err := writer.Write(src)
-	if err != nil {
-		// don't return writer to pool on error.
-		return nil, err
-	}
-
-	// note that the gzip reader must be closed in order for it to write
-	// out trailing contents.  Flush is insufficient.  it is okay to re-use
-	// the writer even after it's closed by Resetting it.
-	err = writer.Close()
-	if err != nil {
-		// don't return writer to pool on error.
-		return nil, err
-	}
+// Name implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Name() string {
+	return "gzip"
+}
 
-	writerPool.Put(writer)
+// NewReader implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) NewReader(r io.Reader) io.ReadCloser {
+	z := readerPool.Get().(*gzip.Reader)
+	z.Reset(r)
+	return &reader{z}
+}
 
-	return buf.Bytes(), err
+// NewWriter implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) NewWriter(w io.Writer) io.WriteCloser {
+	z := writerPool.Get().(*gzip.Writer)
+	z.Reset(w)
+	return &writer{z}
 }
 
-// Decode implements the kafka.CompressionCodec interface.
-func (c CompressionCodec) Decode(src []byte) ([]byte, error) {
-	reader := readerPool.Get().(*gzip.Reader)
-	err := reader.Reset(bytes.NewReader(src))
-	if err != nil {
-		return nil, err
+type reader struct{ *gzip.Reader }
+
+func (r *reader) Close() (err error) {
+	if z := r.Reader; z != nil {
+		r.Reader = nil
+		err = z.Close()
```

Comment:
if `err` is non-nil, should we still return the reader to the pool?  are there any issues we could run into that a reset wouldn't fix?

Reply:
I noticed that this was how the code was constructed in the previous version, I reviewed the standard library code and it looks like Reset really clears everything even when there was an error on the previous read (this is also what's documented). I've also used this construct in other programs before without any issues so I don't anticipate programs with reusing the values after an error. It also prevents cascading failures of blowing up memory usage and increasing GC pressure when error rate increases.